kernel/src/hil/time.rs

```diff
@@ -81,20 +90,96 @@ pub trait Alarm: Time {
 
     /// Returns the value set in [`set_alarm`](#tymethod.set_alarm)
     fn get_alarm(&self) -> u32;
+
+    /// Set the client for interrupt events.
+    fn set_client(&self, client: &'a AlarmClient);
+
+    /// Returns whether this alarm is currently active (will eventually trigger
+    /// a callback if there is a client).
+    fn is_enabled(&self) -> bool;
+
+    /// Enables the alarm using the previously set `tics` for the alarm.
+    ///
+    /// Most implementations should use the default implementation which calls `set_alarm` with the
+    /// value returned by `get_alarm` unless there is a more efficient way to achieve the same
+    /// semantics.
+    fn enable(&self) {
+        self.set_alarm(self.get_alarm())
+    }
+
+    /// Disables the alarm.
+    ///
+    /// The implementation will _always_ disable the alarm, however, it may be possible for an
+    /// alarm to have already expired but the event not delivered to the client. In this case, the
+    /// implementation must return [`FAIL`] letting the caller know that an event for the alarm
+    /// will still be delivered.
+    fn disable(&self) -> ReturnCode;
 }
 
-/// A client of an implementor of the [`Alarm`](trait.Alarm.html) trait.
-pub trait Client {
+/// A client of an implementer of the [`Alarm`](trait.Alarm.html) trait.
+pub trait AlarmClient {
     /// Callback signaled when the alarm's clock reaches the value set in
     /// [`Alarm#set_alarm`](trait.Alarm.html#tymethod.set_alarm).
     fn fired(&self);
 }
 
 /// The `Timer` trait models a timer that can notify when a particular interval
 /// has elapsed.
-pub trait Timer: Time {
+pub trait Timer<'a>: Time {
```

Comment:
Do we have anything in Tock that implemented `Timer`? If not, should we? Or is it really just applications that actually need timers? This is kind of interesting.

Reply:
We have nothing in Tock that currently implements `Timers (part of the issue with settling on an interface, IMO). The NRF52 has timer hardware that’s used in the radio stack, but it uses hardware shorts directly instead of getting timer expiration through interrupts (which is necessary to meet timing constraints).

Arguably it might be useful in the 802.15.4  stack (e.g. for back off or something).

Alternatively, we can view this as a best guess interface for now so there is an incentive to use it when the need arises, and see how well it fits then.